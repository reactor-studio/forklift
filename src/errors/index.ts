export { default as BadRequestError } from './bad-request';
export { default as ForbiddenError } from './forbidden';
export { default as NotFoundError } from './not-found';
export { default as ConflictError } from './conflict';
export { default as ForkliftError } from './forklift-error';
export { default as InputError } from './input-error';
export { default as OutputError } from './output-error';
