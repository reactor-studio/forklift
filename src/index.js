import IO from './io';

export { IoError, ioErrorAdapter } from './io-error';
export { Status } from './status';
export { setLocals, getLocals } from './locals';
export { IO };
