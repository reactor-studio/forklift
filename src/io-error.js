export class IoError extends Error {
  constructor(message) {
    super(message);
    this.name = 'IoError';
    this.message = message;
    this.status = 400;
    this.title = 'Bad request';
    Error.captureStackTrace(this, this.constructor);
  }
}

export class IoResError extends IoError {
  constructor(message) {
    super(message);
    this.name = 'IoResError';
    this.status = 500;
    this.title = 'Response data is invalid';
  }
}

export const ioErrorAdapter = {
  toJson(err) {
    let error = null;
    if (err.name === 'IoError' || err.name === 'IoResError') {
      error = {
        status: err.status,
        title: err.title,
        detail: err.message,
        meta: {
          trace: err,
        },
      };
    }

    return error;
  },
};
