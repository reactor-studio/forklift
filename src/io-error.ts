type Name = 'IoReqError' | 'IoResError';

export interface ErrorDetails {
  why: string;
  where: string;
  how: string;
}

abstract class IoError extends Error {
  readonly name: Name;

  readonly message: string;

  readonly status: number;

  readonly title: string;

  readonly details: object;

  constructor(
    message: string,
    name: Name,
    status: number,
    title: string,
    details?: ErrorDetails,
  ) {
    super(message);
    this.name = name;
    this.message = message;
    this.status = status;
    this.title = title;
    this.details = details;
    Error.captureStackTrace(this, this.constructor);
  }
}

export class IoReqError extends IoError {
  constructor(message: string, details?: ErrorDetails) {
    super(message, 'IoReqError', 400, 'Bad request', details);
  }
}

export class IoResError extends IoError {
  constructor(message: string, details?: ErrorDetails) {
    super(message, 'IoResError', 500, 'Response data is invalid', details);
  }
}

interface IoErrorJson {
  status: number;
  title: string;
  detail: string;
  meta: { trace: object };
}

export const ioErrorAdapter = {
  toJson(err: Error): IoErrorJson {
    let error = null;
    if (err.name === 'IoReqError' || err.name === 'IoResError') {
      error = {
        status: (err as IoError).status,
        title: (err as IoError).title,
        detail: (err as IoError).details,
        meta: {
          trace: err,
        },
      };
    }

    return error;
  },
};
