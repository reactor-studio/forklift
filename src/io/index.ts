export { IO } from './io';
export { Status, statusOptions } from './status';
export { setLocals, getLocals } from './locals';
