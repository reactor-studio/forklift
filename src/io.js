import _ from 'lodash';
import tv4 from 'tv4';
import { Status } from './status';
import { IoError, IoResError } from './io-error';

const jsonContentType = 'application/json';
const supportedContentTypes = [jsonContentType, '*/*', '*'];

export const statusOptions = {
  [Status.OK]: {
    code: 200,
    shouldSerializeData: true,
  },
  [Status.CREATED]: {
    code: 201,
    shouldSerializeData: true,
  },
  [Status.NO_CONTENT]: {
    code: 204,
    shouldSerializeData: false,
  },
  [Status.BAD_REQUEST]: {
    code: 400,
    shouldSerializeData: false,
  },
  [Status.UNAUTHORIZED]: {
    code: 401,
    shouldSerializeData: false,
  },
  [Status.FORBIDDEN]: {
    code: 403,
    shouldSerializeData: false,
  },
  [Status.NOT_FOUND]: {
    code: 404,
    shouldSerializeData: false,
  },
};

export class IO {
  static set(target, data, status = Status.OK, path = null) {
    if (path) {
      _.set(target, `locals.io.data.${path}`, data);
    } else {
      _.set(target, 'locals.io.data', data);
    }
    _.set(target, 'locals.io.status', status);
  }

  static get(target, localsKey = '') {
    const data = _.get(target, 'locals.io.data');
    if (localsKey) {
      return _.get(data, localsKey);
    }

    return data;
  }

  static getStatus(target) {
    return _.get(target, 'locals.io.status');
  }

  /**
   * Sets the IO object with the status CREATED on the
   * given express object (req/res).
   *
   * @param target The express request or response.
   * @param data The IO data to set.
   */
  static setCreated(target, data) {
    IO.set(target, data, Status.CREATED);
  }

  /**
   * Sets an empty response on the given express object (req/res).
   * This method will clear any existing IO data, and set the status
   * to NO_CONTENT.
   *
   * @param target The express request or response.
   */
  static setEmpty(target) {
    IO.set(target, null, Status.NO_CONTENT);
  }

  /**
   * Sets an empty response on the given express object (req/res).
   * This method will clear any existing IO data, and set the status
   * to BAD_REQUEST.
   *
   * @param target The express request or response.
   */
  static setBadRequest(target) {
    IO.set(target, null, Status.BAD_REQUEST);
  }

  /**
   * Sets an empty response on the given express object (req/res).
   * This method will clear any existing IO data, and set the status
   * to UNAUTHORIZED.
   *
   * @param target The express request or response.
   */
  static setUnauthorized(target) {
    IO.set(target, null, Status.UNAUTHORIZED);
  }

  /**
   * Sets an empty response on the given express object (req/res).
   * This method will clear any existing IO data, and set the status
   * to FORBIDDEN.
   *
   * @param target The express request or response.
   */
  static setForbidden(target) {
    IO.set(target, null, Status.FORBIDDEN);
  }

  /**
   * Sets an empty response on the given express object (req/res).
   * This method will clear any existing IO data, and set the status
   * to NOT_FOUND.
   *
   * @param target The express request or response.
   */
  static setNotFound(target) {
    IO.set(target, null, Status.NOT_FOUND);
  }

  static prepareResponse(res) {
    const ioStatus = IO.getStatus(res) || Status.NO_CONTENT;
    const status = statusOptions[ioStatus];
    res.status(status.code);

    return status.shouldSerializeData;
  }

  static setResponseHeaders(res) {
    res.set('Content-Type', jsonContentType);
  }

  static validateResource(resource, schema) {
    const isBodyRespectingSchema = () => tv4.validateResult(resource, schema);
    const errorObject = isBodyRespectingSchema();
    if (errorObject.error) {
      const message = _.get(errorObject, 'error.subErrors')
        ? `It can be any of these errors [${_.map(
            errorObject.error.subErrors,
            subError => subError.message,
          )}]`
        : errorObject.error.message;
      const errorDetails = {
        why: 'Resource does not respect the schema',
        where: `${errorObject.error.dataPath}`,
        how: message,
      };
      return errorDetails;
    }
    return null;
  }

  constructor(reqSchema, options, resSchema) {
    this.reqSchema = reqSchema;
    this.resSchema = resSchema;
    this.options = options;
  }

  validateRequestHeaders(req) {
    if (req.headers['content-type'].indexOf('application/json') < 0) {
      const message = 'Please use application-json as Content-Type';
      throw new IoError(message);
    }

    const acceptHeader = req.get('Accept') || '';
    const contentTypes = [];
    const customContentTypes = _.get(this.options, 'contentTypes', []);
    // Remove (optional) quality value from each accepted type.
    // https://www.w3.org/Protocols/rfc2616/rfc2616-sec3.html#sec3.9
    acceptHeader
      .split(',')
      .forEach(type => contentTypes.push(type.split(';')[0]));

    customContentTypes.forEach(type => contentTypes.push(type));

    const acceptedContentTypes = _.intersection(
      contentTypes,
      supportedContentTypes,
    );

    if (_.isEmpty(acceptedContentTypes)) {
      const message = `${'Client does not accept JSON responses. ' +
        'Did you set the correct "Accept" header?'}${JSON.stringify(
        contentTypes,
      )}`;
      throw new IoError(message);
    }
  }

  validateRequest(req) {
    this.validateRequestHeaders(req);
    if (this.reqSchema) {
      const errorDetails = IO.validateResource(req.body, this.reqSchema);
      if (errorDetails) {
        throw new IoError(errorDetails);
      }
    }
  }

  validateResponse(data) {
    if (this.resSchema) {
      const errorDetails = IO.validateResource(data, this.resSchema);
      if (errorDetails) {
        throw new IoResError(errorDetails);
      }
    }
  }

  processRequest() {
    return (req, res, next) => {
      this.validateRequest(req);
      IO.set(req, req.body);
      next();
    };
  }

  sendResponse() {
    return (req, res, next) => {
      try {
        const shouldSerializeData = IO.prepareResponse(res);
        if (!shouldSerializeData) {
          res.end();
          return null;
        }
        const data = IO.get(res);
        if (_.isEmpty(data)) {
          return next(new IoResError('No data to serialize'));
        }
        this.validateResponse(data);
        res.json(data);
        return null;
      } catch (error) {
        return next(error);
      }
    };
  }
}
