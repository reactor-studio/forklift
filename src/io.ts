import _ from 'lodash';
import tv4 from 'tv4';
import { Request, Response, NextFunction } from 'express';
import { IoReqError, IoResError, ErrorDetails } from './io-error';
import { Status, statusOptions } from './status';

const jsonContentType = 'application/json';
const supportedContentTypes = [jsonContentType, '*/*', '*'];

export class IO {
  static set(
    target: object,
    data: any,
    status: Status = Status.OK,
    path: string = null,
  ): object {
    if (path) {
      _.set(target, `locals.io.data.${path}`, data);
    } else {
      _.set(target, 'locals.io.data', data);
    }
    return _.set(target, 'locals.io.status', status);
  }

  static get(target: object, localsKey = ''): any {
    const data = _.get(target, 'locals.io.data');
    if (localsKey) {
      return _.get(data, localsKey);
    }

    return data;
  }

  static getStatus(target: object): any {
    return _.get(target, 'locals.io.status');
  }

  /**
   * Sets the IO object with the status CREATED on the
   * given express object (req/res).
   *
   * @param target The express request or response.
   * @param data The IO data to set.
   */
  static setCreated(target: object, data: any): object {
    return IO.set(target, data, Status.CREATED);
  }

  /**
   * Sets an empty response on the given express object (req/res).
   * This method will clear any existing IO data, and set the status
   * to NO_CONTENT.
   *
   * @param target The express request or response.
   */
  static setEmpty(target: object): object {
    return IO.set(target, null, Status.NO_CONTENT);
  }

  /**
   * Sets an empty response on the given express object (req/res).
   * This method will clear any existing IO data, and set the status
   * to BAD_REQUEST.
   *
   * @param target The express request or response.
   */
  static setBadRequest(target: object): object {
    return IO.set(target, null, Status.BAD_REQUEST);
  }

  /**
   * Sets an empty response on the given express object (req/res).
   * This method will clear any existing IO data, and set the status
   * to UNAUTHORIZED.
   *
   * @param target The express request or response.
   */
  static setUnauthorized(target: object): object {
    return IO.set(target, null, Status.UNAUTHORIZED);
  }

  /**
   * Sets an empty response on the given express object (req/res).
   * This method will clear any existing IO data, and set the status
   * to FORBIDDEN.
   *
   * @param target The express request or response.
   */
  static setForbidden(target: object): object {
    return IO.set(target, null, Status.FORBIDDEN);
  }

  /**
   * Sets an empty response on the given express object (req/res).
   * This method will clear any existing IO data, and set the status
   * to NOT_FOUND.
   *
   * @param target The express request or response.
   */
  static setNotFound(target: object): object {
    return IO.set(target, null, Status.NOT_FOUND);
  }

  static prepareResponse(res: Response): boolean {
    const ioStatus = IO.getStatus(res) || Status.NO_CONTENT;
    const status = statusOptions[ioStatus];
    res.status(status.code);

    return status.shouldSerializeData;
  }

  static setResponseHeaders(res: Response): Response {
    return res.set('Content-Type', jsonContentType);
  }

  static validateResource(resource: any, schema: tv4.JsonSchema): ErrorDetails {
    const isBodyRespectingSchema = (): tv4.SingleResult =>
      tv4.validateResult(resource, schema);
    const errorObject = isBodyRespectingSchema();
    if (errorObject.error) {
      const message = _.get(errorObject, 'error.subErrors')
        ? `It can be any of these errors [${_.map(
            errorObject.error.subErrors,
            subError => subError.message,
          )}]`
        : errorObject.error.message;
      const errorDetails = {
        why: 'Resource does not respect the schema',
        where: `${errorObject.error.dataPath}`,
        how: message,
      };
      return errorDetails;
    }
    return null;
  }

  private reqSchema: tv4.JsonSchema;

  private resSchema: tv4.JsonSchema;

  private options: object;

  constructor(
    reqSchema: tv4.JsonSchema,
    options: object,
    resSchema: tv4.JsonSchema,
  ) {
    this.reqSchema = reqSchema;
    this.resSchema = resSchema;
    this.options = options;
  }

  private validateRequestHeaders(req: Request): void {
    if (req.headers['content-type'].indexOf('application/json') < 0) {
      const message = 'Please use application-json as Content-Type';
      throw new IoReqError(message);
    }

    const acceptHeader = req.get('Accept') || '';
    const contentTypes = [];
    const customContentTypes = _.get(this.options, 'contentTypes', []);
    // Remove (optional) quality value from each accepted type.
    // https://www.w3.org/Protocols/rfc2616/rfc2616-sec3.html#sec3.9
    acceptHeader
      .split(',')
      .forEach(type => contentTypes.push(type.split(';')[0]));

    customContentTypes.forEach(type => contentTypes.push(type));

    const acceptedContentTypes = _.intersection(
      contentTypes,
      supportedContentTypes,
    );

    if (_.isEmpty(acceptedContentTypes)) {
      const message = `${'Client does not accept JSON responses. ' +
        'Did you set the correct "Accept" header?'}${JSON.stringify(
        contentTypes,
      )}`;
      throw new IoReqError(message);
    }
  }

  private validateRequest(req: Request): void {
    this.validateRequestHeaders(req);
    if (this.reqSchema) {
      const errorDetails = IO.validateResource(req.body, this.reqSchema);
      if (errorDetails) {
        throw new IoReqError(errorDetails.why, errorDetails);
      }
    }
  }

  private validateResponse(data: object): void {
    if (this.resSchema) {
      const errorDetails = IO.validateResource(data, this.resSchema);
      if (errorDetails) {
        throw new IoResError(errorDetails.why, errorDetails);
      }
    }
  }

  processRequest() {
    return (req: Request, _res: Response, next: NextFunction): void => {
      this.validateRequest(req);
      IO.set(req, req.body);
      next();
    };
  }

  sendResponse() {
    return (_req: Request, res: Response, next: NextFunction): void => {
      try {
        const shouldSerializeData = IO.prepareResponse(res);
        if (!shouldSerializeData) {
          res.end();
          return null;
        }
        const data = IO.get(res);
        if (_.isEmpty(data)) {
          return next(new IoResError('No data to serialize'));
        }
        this.validateResponse(data);
        res.json(data);
        return null;
      } catch (error) {
        return next(error);
      }
    };
  }
}
