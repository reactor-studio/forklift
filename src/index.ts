export { IoReqError, IoResError, ioErrorAdapter } from './io-error';
export { Status, statusOptions } from './status';
export { setLocals, getLocals } from './locals';
import { IO } from './io';

export default IO;
