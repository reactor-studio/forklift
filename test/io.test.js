import lodash from 'lodash';
import { Response } from 'jest-express/lib/response';
import { Request } from 'jest-express/lib/request';
import { IO, Status } from '../src';

describe('io static function', () => {
  test('set sets provided data', () => {
    const target = { test: true };
    IO.set(target, { another: false });
    expect(lodash.get(target, 'locals.io.data')).toEqual({ another: false });
  });

  test('set sets provided status', () => {
    const target = { test: true };
    IO.set(target, {}, Status.NOT_FOUND);
    expect(lodash.get(target, 'locals.io.status')).toBe('not-found');
  });

  test('set sets default status to ok', () => {
    const target = { test: true };
    IO.set(target, {});
    expect(lodash.get(target, 'locals.io.status')).toBe('ok');
  });

  test('get returns the right value', () => {
    const target = {};
    lodash.set(target, 'locals.io.data', { data: true });
    expect(IO.get(target)).toEqual({ data: true });
  });

  test('get with local key returns data within local key', () => {
    const target = {};
    lodash.set(target, 'locals.io.data.some.key', { data: true });
    expect(IO.get(target, 'some.key')).toEqual({ data: true });
  });

  test('getStatus returns status', () => {
    const target = {};
    lodash.set(target, 'locals.io.status', 'ok');
    expect(IO.getStatus(target)).toBe('ok');
  });

  test('setCreated sets status created', () => {
    const target = {};
    IO.setCreated(target, 'data');
    expect(lodash.get(target, 'locals.io.status')).toBe('created');
  });

  test('setEmpty clears data and sets status no-content', () => {
    const target = {};
    lodash.set(target, 'locals.io.data', { data: true });
    IO.setEmpty(target);
    expect(lodash.get(target, 'locals.io.data')).toBeNull();
    expect(lodash.get(target, 'locals.io.status')).toBe('no-content');
  });

  test('setBadRequest clears data & sets bad-request status', () => {
    const target = {};
    lodash.set(target, 'locals.io.data', { data: true });
    IO.setBadRequest(target);
    expect(lodash.get(target, 'locals.io.data')).toBeNull();
    expect(lodash.get(target, 'locals.io.status')).toBe('bad-request');
  });

  test('setUnauthorized clears data & sets unauthorized status', () => {
    const target = {};
    lodash.set(target, 'locals.io.data', { data: true });
    IO.setUnauthorized(target);
    expect(lodash.get(target, 'locals.io.data')).toBeNull();
    expect(lodash.get(target, 'locals.io.status')).toBe('unauthorized');
  });

  test('setForbidden clears data & sets forbidden status', () => {
    const target = {};
    lodash.set(target, 'locals.io.data', { data: true });
    IO.setForbidden(target);
    expect(lodash.get(target, 'locals.io.data')).toBeNull();
    expect(lodash.get(target, 'locals.io.status')).toBe('forbidden');
  });

  test('setNotFound clears data & sets not-found status', () => {
    const target = {};
    lodash.set(target, 'locals.io.data', { data: true });
    IO.setNotFound(target);
    expect(lodash.get(target, 'locals.io.data')).toBeNull();
    expect(lodash.get(target, 'locals.io.status')).toBe('not-found');
  });
});

describe('io function', () => {
  let io;
  let req;
  let res;

  const simpleSchema = {
    type: 'object',
    properties: {
      firstField: {
        type: 'string',
        description: "The object' first field.",
      },
      secondField: {
        type: 'string',
        description: "The object's second field.",
      },
    },
    required: ['firstField', 'secondField'],
  };

  beforeEach(() => {
    req = new Request();
    res = new Response();
    io = new IO(simpleSchema, { options: 1 });
  });

  afterEach(() => {
    req.resetMocked();
    res.resetMocked();
  });

  test('prepareResponse sets correct status code', () => {
    lodash.set(res, 'locals.io.status', 'ok');
    IO.prepareResponse(res);
    expect(res.status).toHaveBeenCalledWith(200);
  });

  test('setResponseHeaders sets application/json to content-type', () => {
    IO.setResponseHeaders(res);
    expect(res.getHeader('Content-Type')).toEqual('application/json');
  });

  test('constructor sets correct schema', () => {
    expect(io.schema).toEqual(simpleSchema);
  });

  test('constructor sets correct options', () => {
    expect(io.options).toEqual({ options: 1 });
  });

  test('validateRequestHeaders throws on empty accept header', () => {
    req.headers['content-type'] = 'application/json';
    expect(() =>
      io.validateRequestHeaders(req),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Client does not accept JSON responses. Did you set the correct \\"Accept\\" header?[\\"\\"]"`,
    );
  });

  test('validateResource throws on request not respecting schema', () => {
    req.setBody({ firstField: 'value', secondBADField: 'value' });
    expect(() => io.validateResource(req)).toThrowErrorMatchingSnapshot();
  });

  test('validateResource does not throw on request respecting schema', () => {
    req.setBody({ firstField: 'value', secondField: 'value' });
    expect(() => io.validateResource(req)).not.toThrow();
  });

  test('processRequest throws on wrong content-type', () => {
    const next = err => {
      throw err;
    };
    req.setHeaders('content-type', 'text/plain');
    expect(() =>
      io.processRequest()(req, res, next),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Please use application-json as Content-Type"`,
    );
  });

  test('sendResponse ends response on non-serializable data', () => {
    IO.set(res, {}, null, Status.NOT_FOUND);
    io.sendResponse()(req, res, null);
    expect(res.end).toHaveBeenCalled();
  });

  test('sendResponse throws on empty data to be serialized', () => {
    const next = err => {
      throw err;
    };
    IO.set(res, null, Status.OK);
    expect(() =>
      io.sendResponse()(req, res, next),
    ).toThrowErrorMatchingInlineSnapshot(`"No data to serialize"`);
  });

  test('sendResponse sends json response', () => {
    IO.set(res, { test: 'test!' });
    io.sendResponse()(req, res, null);
    expect(res.json).toHaveBeenCalled();
  });
});
