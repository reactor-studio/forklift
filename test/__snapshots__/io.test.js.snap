// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`io function validateResource throws on request not respecting schema 1`] = `
Object {
  "how": "Missing required property: secondField",
  "where": "request/body",
  "why": "Body does not respect the schema",
}
`;
